### PR TITLE
fix: add generic parameters in convert to manual impl assist

### LINKED ISSUE #10041: Convert to manual impl should add generic parameters
manul

Code & tests in question:  https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide_assists/src/handlers/replace_derive_with_manual_impl.rs

### PR DESCRIPTION
**Pull request by TheDoctor314**
Fixes #10041.

### CODE REVIEW

File: crates/syntax/src/ast/make.rs [hunk 1]
```diff
@@ -149,8 +149,13 @@ pub fn impl_(
...
 
-pub fn impl_trait(trait_: ast::Path, ty: ast::Path) -> ast::Impl {
-    ast_from_text(&format!("impl {} for {} {{}}", trait_, ty))
+pub fn impl_trait(
+    trait_: ast::Path,
+    ty: ast::Path,
+    ty_params: Option<ast::GenericParamList>,
+) -> ast::Impl {
+    let ty_params = ty_params.map_or_else(String::new, |params| params.to_string());
+    ast_from_text(&format!("impl{2} {} for {}{2} {{}}", trait_, ty, ty_params))
```
Comment (Veykril) [inline crates/syntax/src/ast/make.rs:158]: Slightly incorrect, `ast::GenericParamList` also contains bounds which are invalid in the second use of the `ty_params` here in the formatting string. See `generate_impl_text` in `crates/ide_assists/src/utils.rs` which also does something similar(might be possible to reuse that function here as well somehow, as its already used in other branches in this assist)

Comment (TheDoctor314) [reply crates/syntax/src/ast/make.rs:158]: Should I just copy the code from `generate_trait_impl_text` because there is no way to use that function here?

Comment (Veykril) [reply crates/syntax/src/ast/make.rs:158]: If you can't make use of it as is, then ye just copy the things you need.

File: crates/syntax/src/ast/make.rs [hunk 2]
```diff
@@ -149,8 +153,65 @@ pub fn impl_(
     ast_from_text(&format!("impl{} {}{} {{}}", params, ty, ty_params))
 }
 
-pub fn impl_trait(trait_: ast::Path, ty: ast::Path) -> ast::Impl {
-    ast_from_text(&format!("impl {} for {} {{}}", trait_, ty))
+pub fn impl_trait(trait_: &ast::Path, adt: &ast::Adt, code: &str) -> ast::Impl {
```
Comment (Veykril) [inline crates/syntax/src/ast/make.rs:156]: Ah I think I misunderstood you here a bit. The `make` module itself shouldn't do anything fancy aside from stupidly creating syntax trees.

So the signature of this function here should be something like
```rs
fn impl_trait(generics: ast::GenericParamList, trait_: Option<ast::Path>, target: ast::Type, assoc_item_list: ast::AssocItemList) -> ast::Impl
```
and it should only do minimal work to create the `ast::Impl` node, so the caller will have to do a bit more work in creating the input nodes via `make` as well then.

Comment (TheDoctor314) [reply crates/syntax/src/ast/make.rs]: I don't think that signature is right, because the generics should be optional?

**Pull request merged**